oak_functions_launcher/tests/integration_test.rs

```diff
@@ -53,22 +54,134 @@ async fn test_launcher_virtual() {
     let wasm_path = oak_functions_test_utils::build_rust_crate_wasm("key_value_lookup")
         .expect("Failed to build Wasm module");
 
+    let port = portpicker::pick_unused_port().expect("failed to pick a port");
+
     let _background = xtask::testing::run_background(
-        xtask::launcher::run_oak_functions_launcher_example(&variant, &wasm_path),
+        xtask::launcher::run_oak_functions_launcher_example(&variant, &wasm_path, port),
     )
     .await;
 
     // Wait for the server to start up.
     tokio::time::sleep(Duration::from_secs(20)).await;
 
-    let mut client = oak_functions_client::OakFunctionsClient::new("http://localhost:8080")
-        .await
-        .unwrap();
+    let mut client =
+        oak_functions_client::OakFunctionsClient::new(&format!("http://localhost:{port}"))
+            .await
+            .unwrap();
 
     let response = client.invoke(b"test_key").await.expect("failed to invoke");
     assert_eq!(response, b"test_value");
 }
 
+// Allow enough worker threads to collect output from background tasks.
+#[tokio::test(flavor = "multi_thread", worker_threads = 3)]
+async fn test_launcher_echo_virtual() {
+    if xtask::testing::skip_test() {
+        log::info!("skipping test");
+        return;
+    }
+
+    xtask::testing::run_step(xtask::launcher::build_stage0()).await;
+    xtask::testing::run_step(xtask::launcher::build_binary(
+        "build Oak Restricted Kernel binary",
+        xtask::launcher::OAK_RESTRICTED_KERNEL_BIN_DIR
+            .to_str()
+            .unwrap(),
+    ))
+    .await;
+    let variant = xtask::launcher::LauncherMode::Virtual("oak_functions_enclave_app".to_string());
+    xtask::testing::run_step(xtask::launcher::build_binary(
+        "build Oak Functions enclave app",
+        &variant.enclave_crate_path(),
+    ))
+    .await;
+
+    let wasm_path = oak_functions_test_utils::build_rust_crate_wasm("echo")
+        .expect("Failed to build Wasm module");
+
+    let port = portpicker::pick_unused_port().expect("failed to pick a port");
+
+    let _background = xtask::testing::run_background(
+        xtask::launcher::run_oak_functions_launcher_example(&variant, &wasm_path, port),
+    )
+    .await;
+
+    // Wait for the server to start up.
+    tokio::time::sleep(Duration::from_secs(20)).await;
```

Comment:
Does not this make the test flaky? Is it possible to implement a message passing mechanism? (in a future PR perhaps)

Reply:
I thought maybe `run_oak_functions_launcher_example_with_lookup_data` or similar methods that run in the background could somehow signal that they are completed, but I have not looked at the details of these functions, so it might not be feasible.